### DISCUSSION TITLE
Send statistics to a remote MySQL server?

### DISCUSSION THREAD

**Discussion by oxygenic** [arendst/Tasmota#16877] [Q&A]
I'm using a Wifi wall-plug with power measurement and would like to have some more detailled statistics about daily/weekly/monthly power usage. As the statistics in Tasmota are extremely rudimentary, it seems I have to do that somehow outside.

My wall-plug is located behind a firewall, so it can access an external server but the server can not access the wall-plug.

What I want to do is to write the "Energy yesterday" into a MySQL database once a day.

My question: how can I do that exactly? Means how can I access this value, how can I send it to an external server and how can I trigger this function once a day?

The server itself e.g. could provide a PHP-script that accepts the energy value as parameter, but the question is, how can I get the value and how can I send it (HTTP GET?) to the server?

**Comment by arendst** [comment]
Nice excersie.

Why not use already available solutions like a local server (raspberry Pi - low power) running Domoticz resulting in this:

![image](https://user-images.githubusercontent.com/11044339/197343498-aa01b14f-bf57-43ab-9ba5-83f0e8c486c5.png)

The above is the result of a Sonoff Pow monitoring e-bike battery charging over the years.

**Comment by oxygenic** [comment]
> Why not use already available solutions like a local server (raspberry Pi - low power)

Because I do not have an local server, but a remote one. So why having an other piece of electronic while there is already one available?

**Reply by joba-1** [reply]
there are a lot if options…
* is mysql a requirement? Tasmota has a InfluxDB interface.
* You could send syslog messages (tele messages)
* rule based websend
…

**Reply by oxygenic** [reply]
No, it is not a requirement - I'm open to everything that can work in my usage scenario.

So...how is your idea going to work exactly?

**Reply by joba-1** [reply]
first one more question: you write you don't have a local server. Does that mean tasmota is connected to a remote mqtt broker? Subscribing to that broker would be the cleanest solution.
… if you don’t want to compile tasmota yourself - I think there is no precompiled tasmota with InfluxDB

**Reply by oxygenic** [reply]
At the moment I'm at the very beginning, means I have nothing than this wall-plug with Tasmota (standard version, nothing self-compiled) and the remote server that is intended to receive my "Yesterday energy" data.

They are not yet connected in any way as this is exactly what I want to find out how to do that.

**Reply by sfromis** [reply]
If you do not want statistics in general, but only numbers once a day, it would be possible to create rules to do a http (not https) request with the data. Of course, opening *anything* to unencrypted Internet cannot be recommended. Tasmota also has a Telegram interface if you trust the Telegram platform.

**Reply by joba-1** [reply]
ok, I think you need nothing special. The normal setup for tasmota would be to connect it to an mqtt broker. The broker can run where you wanted to run your mysql server. Tasmota sends status infos to the broker and can receive commands from the broker. I use mosquitto as my broker.
Home automation software like e.g. domoticz (that can also run on the remote server) then subscribes at the broker to these messages and saves them in a database. That is what Theo did.

**Reply by oxygenic** [reply]
> If you do not want statistics in general, but only numbers once a day, it would be possible to create rules to do a http (not https) request with the data. Of course, opening _anything_ to unencrypted Internet cannot be recommended. Tasmota also has a Telegram interface if you trust the Telegram platform.

Yes, that's exactly what I'm looking for. So: how can I do that? I already digged through the docs but couldn't even find out how to read that "Yesterday energy" value

**Reply by sfromis** [reply]
[Rules](https://tasmota.github.io/docs/Rules/) can do a lot, but takes a bit of time to grok and get used to.

**Reply by barbudor** [reply]
Rules : https://tasmota.github.io/docs/Rules/
WebQuery : https://tasmota.github.io/docs/Commands/#webquery

You will need 2 rules:
- 1 triggered at each teleperiod SENSOR message to continuously save the value of EnergyYesterday into a variable
- 1 triggered once a day, for example 10 minutes after midnight, to send the GET request to your PHP Script (Over HTTP, not https)

**Reply by oxygenic** [reply]
This sounds good, but unfortunately the documentation does not seem to reflect this (or I can't find it):

> * 1 triggered at each teleperiod SENSOR message to continuously save the value of EnergyYesterday into a variable

There are a lot of "teleperiod" messages described, but none that is named "SENSOR" (and to be honest, I do not understand what "SENSOR" has to do with "Energy Yesterday" which itself also isn't mentioned anywhere in this document)

> * 1 triggered once a day, for example 10 minutes after midnight, to send the GET request to your PHP Script (Over HTTP, not https)

This is clear at least parially. So my command would be something like

"WebQuery http://myserver.tld/energy.php?power=<yesterday> GET"

Questions here:

1. Assumed I have solved my first problem, how do I create a GET-string which contains the yesterdays value in <yesterday>?

2. In configuration "Timer" I can choose "Rule" as my "Action" to be done - but I can't select which rule - so how does Tasmota know I want to execute the WebQuery-rule but not the "teleperiod"-rule?

Thanks!

**Reply by barbudor** [reply]
You haven't read enough

> There are a lot of "teleperiod" messages described, but none that is named "SENSOR"

There are only 2 teleperiod messages:
- STATE which provides more "system" information like memory use, estimated load etc...
- SENSOR which provides all measurements from all sensors, including energy sensors

You can see those messages in the console

For the rule, it doesn't matter if the data you are looking for is in STATE or SENSOR, only the key in the json matters.
The key you are looking for is obviously `energy#yesterday`

For a rule triggered from a timer, search in the rule page about `clock#timer=`

Everything is in the docs

**Reply by oxygenic** [reply]
Sorry, but for me all the docs do not make much sense.

There it e.g. is stated: "To trigger only at TelePeriod time, prefix the sensor with the word Tele-"

OK, but I do not want to read any hardware-sensor, I want to read the yesterdays energy value! Here the documentation is talking about sensors, this would make sense when I would like to read the current energy directly from the related sensor, but this is not what I plan to do.

And when I search through the two links you gave to me, there nowhere the tag "energy#yesterday" can be found - so where is this described exactly?

Sorry to say, but I'm really surprised how complicated it is to read just one single value and to send it to the outer world. Is it possible to understand this without being a hardware-hacker/a core-developer of the Tasmota stuff?

By any chance, is it possible to get just one simple example how reading the value and sending it to a URL could look like?

**Reply by oxygenic** [reply]
> You can see those messages in the console

There are no such messages. All I can see in the console, is a text

`00:00:00.050 CFG: Loaded from flash at F5, Count 145`

No matter what I try to type there, (telperiod, sensor, state, whatever), nothing happens.

**Reply by pkkrusty** [reply]
Type in `weblog 3` so you can see more info in the console. Then you'll see all the status and sensor messages, which will help you in your journey of understanding how rules work. It's like learning a new language. Rules have to be generic enough to fit almost all situations, thus can't be specific enough to be intuitive for a non-native speaker. There are basic like "sentence construction" that take time to get used to like `on energy#yesterday do var1 %value% endon on time#minute=360 do WebQuery http://192.168.1.23:8086/write?db=homeautomation&rp=four_weeks POST ping,device=living-room-speakers value=%var1% endon` Even that example probably has a few syntax errors because I'm writing it while on the can.

**Comment by Noschvie** [comment]
@oxygenic : You should have STATE and SENSOR messages like in the example below in your console output, every 5 minutes.
What are logged in your console ?

```
17:24:41.281 MQT: tele/tasmota_012345/STATE = {"Time":"2022-10-23T17:24:41","Uptime":"140T22:06:03","UptimeSec":12175563,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":5,"POWER1":"ON","POWER2":"ON","Wifi":{"AP":1,"SSId":"my-SSid","BSSId":"34:FC:B9:B7:14:80","Channel":6,"Mode":"11n","RSSI":92,"Signal":-54,"LinkCount":1,"Downtime":"0T00:00:03"}}
17:24:41.292 MQT: tele/tasmota_012345/SENSOR = {"Time":"2022-10-23T17:24:41","ENERGY":{"TotalStartTime":"2020-06-15T19:54:14","Total":51.134,"Yesterday":0.034,"Today":0.024,"Period": 0,"Power": 1,"ApparentPower": 9,"ReactivePower": 9,"Factor":0.16,"Voltage":237,"Current":0.036}}
```

**Reply by barbudor** [reply]
You are using a hardware sensor. How do you think Tasmota can get the energy measurements?

Do you have energy displayed on the web GUI?
If not then your plug is not well configured and you must apply the proper template and activate it.
If yes, you should have this 2 messages as shown by noschvie. Unless you have messed up Teleperiod and set it to 0. Set `Teleperiod 300` to get messages every 5 min.

Teleperiod command here: https://tasmota.github.io/docs/Commands/#teleperiod

**Reply by oxygenic** [reply]
May be Tasmota is internally using a hardware sensor, but I do not want to send the current energy (which for sure comes from a hardware sensor), but the concatenated "**Energy yesterday**" value which is shown in the GUI! So no actual hardware sensor measurement but the last days sum of energy:

[https://i.ibb.co/zP4DQ5j/Unbenannt.png](https://i.ibb.co/zP4DQ5j/Unbenannt.png)

And: the console is empty, there nothing more can be seen than this (this was the case from the very beginning, setting some other teleperiod does not change anything):

[https://i.postimg.cc/cHf0Hn0Y/Unbenannt.png](https://i.postimg.cc/cHf0Hn0Y/Unbenannt.png)

**Reply by Noschvie** [reply]
Reboot the device and provide the console output, I can't belive that there is just one line logged.

**Reply by oxygenic** [reply]
A reboot does not change anything, still only that single line: [https://i.ibb.co/cLfCBSb/Unbenannt.png](https://i.ibb.co/cLfCBSb/Unbenannt.png)

**Reply by Noschvie** [reply]
you should start with a clean re-flash of the device. Until you have more logs...

**Reply by barbudor** [reply]
May be you have disabled the logs with `Weblog 0` ?
Set to `Weblog 2`
If not, your device is badly flash but if you get all the energy details on the main web page, then it should be working
But as I already told you, the trigger to get Energy Yesterday in a rule in `Energy#Yesterday`
This is Tasmota how will extract the Yesterday information from the SENSOR message that should looks like that:
```
17:45:40.619 MQT: tele/skt-barbuxps/SENSOR = {"Time":"2022-10-24T17:45:40+02:00","ENERGY":{"TotalStartTime":"2021-05-28T13:47:56","Total":60.737,"Yesterday":0.278,"Today":0.031,"Period":0.0,"Power":0.0,"ApparentPower":0.0,"ReactivePower":0.0,"Factor":0.00,"Voltage":242,"Current":0.000}} (retained)
```
As you can see, the SENSOR message is where Tasmota has the Energy details, including, but not limited to, Yesterday

**Reply by oxygenic** [reply]
OK, and where can I find and how can I use this "Energy#Yesterday"???

Sorry to say, but I have the feeling nobody really understands this rule-stuff. Asking a clear and unique question does not result in a clear answer or in an example that I can use as starting point, but only in some diffuse "read the docs"-feedback.
And when I have a look into the linked docs, I can not find anything that is mentioned here - such as this "Energy#Yesterday".

I give up, It really does not worth to waste that much time with this stuff nobody seem to really understand. Every cheap Chinese crap works better from the very beginning and has the statistics I'm looking for already integrated.

**Comment by jkwim** [comment]
The way I see is that your question does not explain your device, simply whether it runs tasmota at all.
If you are already running Tasmota then post a sample of your web console output so that others can understand what the messages look like.
